modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/pruning/PartitionPruningPredicate.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * Applies partition pruning to the given colocation group. This group should have the same number of assignments as the source table.
+     *
+     * @param colocationGroup Colocation group.
+     *
+     * @return New colocation group.
+     */
+    public ColocationGroup prunePartitions(ColocationGroup colocationGroup) {
+        assert tablePartitions == colocationGroup.assignments().size() : "Number of partitions does not match";
+
+        if (remainingPartitions == null) {
+            return colocationGroup;
+        }
+
+        Map<String, List<PartitionWithConsistencyToken>> partitionsPerNode = newHashMap(colocationGroup.nodeNames().size());
+        Set<String> newNodes = new HashSet<>();
+
+        for (String nodeName : colocationGroup.nodeNames()) {
+            List<PartitionWithConsistencyToken> partsWithConsistencyTokens = new ArrayList<>();
+
+            for (int p = 0; p < colocationGroup.assignments().size(); p++) {
+                NodeWithConsistencyToken nodeWithConsistencyToken = colocationGroup.assignments().get(p);
+                if (!remainingPartitions.contains(p)) {
+                    continue;
+                }
+
+                if (Objects.equals(nodeName, nodeWithConsistencyToken.name())) {
+                    long t = nodeWithConsistencyToken.enlistmentConsistencyToken();
+
+                    partsWithConsistencyTokens.add(new PartitionWithConsistencyToken(p, t));
+                    newNodes.add(nodeName);
+                }
+            }
+
+            if (!partsWithConsistencyTokens.isEmpty()) {
+                partitionsPerNode.put(nodeName, partsWithConsistencyTokens);
+            }
+        }
+
+        // Keep assignments intact, because they are used by DestinationFactory.
+        return new ColocationGroup(
+                colocationGroup.sourceIds(),
+                List.copyOf(newNodes),
+                colocationGroup.assignments(),
+                partitionsPerNode
+        );
+    }
+
+    @Nullable
+    private static IntSet computeRemainingPartitions(
+            IgniteTable table,
+            PartitionPruningColumns pruningColumns,
+            Object[] dynamicParameters
+    ) {
+        ImmutableIntList keys = table.distribution().getKeys();
+        IntSet remainingPartitions = new IntArraySet(pruningColumns.columns().size());
+        PartitionCalculator partitionCalculator = table.partitionCalculator().get();
+
+        for (Int2ObjectMap<RexNode> columns : pruningColumns.columns()) {
+            for (int key : keys) {
+                RexNode node = columns.get(key);
+                NativeType physicalType = table.descriptor().columnDescriptor(key).physicalType();
+
+                // TODO: https://issues.apache.org/jira/browse/IGNITE-21543
+                //  Remove after this issue makes it possible to have CAST('uuid_str' AS UUID) as value.
+                if (physicalType.spec() == NativeTypeSpec.UUID) {
```

Comment:
uuid still can be passed as dynamic param, right?

Reply:
Fixed.